File: C#/ex35/binary_heap_solution.cs
/**
 * Exercise 35: Binary Heap (Priority Queue)
 * Difficulty: ***** (5/5)
 * 
 * Solution using List<int> to implement a min-heap with standard heap operations.
 */

using System;
using System.Collections.Generic;

class MinHeap
{
    private List<int> heap;

    public MinHeap()
    {
        heap = new List<int>();
    }

    // Helper method to get parent index
    private int Parent(int index)
    {
        return (index - 1) / 2;
    }

    // Helper method to get left child index
    private int LeftChild(int index)
    {
        return 2 * index + 1;
    }

    // Helper method to get right child index
    private int RightChild(int index)
    {
        return 2 * index + 2;
    }

    // Swap elements at two indices
    private void Swap(int i, int j)
    {
        int temp = heap[i];
        heap[i] = heap[j];
        heap[j] = temp;
    }

    // Bubble up element to maintain heap property
    private void HeapifyUp(int index)
    {
        while (index > 0 && heap[Parent(index)] > heap[index])
        {
            Swap(index, Parent(index));
            index = Parent(index);
        }
    }

    // Bubble down element to maintain heap property
    private void HeapifyDown(int index)
    {
        int minIndex = index;
        int left = LeftChild(index);
        int right = RightChild(index);

        if (left < heap.Count && heap[left] < heap[minIndex])
        {
            minIndex = left;
        }

        if (right < heap.Count && heap[right] < heap[minIndex])
        {
            minIndex = right;
        }

        if (index != minIndex)
        {
            Swap(index, minIndex);
            HeapifyDown(minIndex);
        }
    }

    // Insert element into heap
    public void Insert(int value)
    {
        heap.Add(value);
        HeapifyUp(heap.Count - 1);
    }

    // Extract minimum element
    public int? ExtractMin()
    {
        if (heap.Count == 0)
        {
            return null;
        }

        int min = heap[0];
        heap[0] = heap[heap.Count - 1];
        heap.RemoveAt(heap.Count - 1);

        if (heap.Count > 0)
        {
            HeapifyDown(0);
        }

        return min;
    }

    // Peek at minimum element without removing
    public int? Peek()
    {
        if (heap.Count == 0)
        {
            return null;
        }
        return heap[0];
    }

    // Check if heap is empty
    public bool IsEmpty()
    {
        return heap.Count == 0;
    }
}

class Program
{
    static void Main()
    {
        int n = int.Parse(Console.ReadLine());
        MinHeap heap = new MinHeap();

        for (int i = 0; i < n; i++)
        {
            string[] input = Console.ReadLine().Split();
            string operation = input[0];

            if (operation == "insert")
            {
                int value = int.Parse(input[1]);
                heap.Insert(value);
            }
            else if (operation == "extract")
            {
                int? min = heap.ExtractMin();
                if (min.HasValue)
                {
                    Console.WriteLine(min.Value);
                }
                else
                {
                    Console.WriteLine("Heap is empty");
                }
            }
            else if (operation == "peek")
            {
                int? min = heap.Peek();
                if (min.HasValue)
                {
                    Console.WriteLine(min.Value);
                }
                else
                {
                    Console.WriteLine("Heap is empty");
                }
            }
            else if (operation == "empty")
            {
                Console.WriteLine(heap.IsEmpty().ToString().ToLower());
            }
        }
    }
}

File: C/ex44/binary_heap_solution.c
/**
 * Exercise 44: Binary Min-Heap
 * Difficulty: ***** (5/5)
 * 
 * Solution implementing a Min-Heap with array-based representation.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_SIZE 100000

typedef struct {
    int data[MAX_SIZE];
    int size;
} MinHeap;

void initHeap(MinHeap *heap) {
    heap->size = 0;
}

int parent(int i) {
    return (i - 1) / 2;
}

int leftChild(int i) {
    return 2 * i + 1;
}

int rightChild(int i) {
    return 2 * i + 2;
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void heapifyUp(MinHeap *heap, int index) {
    while (index > 0 && heap->data[parent(index)] > heap->data[index]) {
        swap(&heap->data[index], &heap->data[parent(index)]);
        index = parent(index);
    }
}

void heapifyDown(MinHeap *heap, int index) {
    int minIndex = index;
    int left = leftChild(index);
    int right = rightChild(index);
    
    if (left < heap->size && heap->data[left] < heap->data[minIndex]) {
        minIndex = left;
    }
    
    if (right < heap->size && heap->data[right] < heap->data[minIndex]) {
        minIndex = right;
    }
    
    if (index != minIndex) {
        swap(&heap->data[index], &heap->data[minIndex]);
        heapifyDown(heap, minIndex);
    }
}

void insert(MinHeap *heap, int value) {
    if (heap->size >= MAX_SIZE) {
        return; // Heap is full
    }
    
    heap->data[heap->size] = value;
    heapifyUp(heap, heap->size);
    heap->size++;
}

int extractMin(MinHeap *heap, int *result) {
    if (heap->size == 0) {
        return 0; // Empty heap
    }
    
    *result = heap->data[0];
    heap->data[0] = heap->data[heap->size - 1];
    heap->size--;
    
    if (heap->size > 0) {
        heapifyDown(heap, 0);
    }
    
    return 1;
}

int peek(MinHeap *heap, int *result) {
    if (heap->size == 0) {
        return 0; // Empty heap
    }
    
    *result = heap->data[0];
    return 1;
}

int isEmpty(MinHeap *heap) {
    return heap->size == 0;
}

int main() {
    int n;
    scanf("%d", &n);
    getchar();
    
    MinHeap heap;
    initHeap(&heap);
    
    for (int i = 0; i < n; i++) {
        char operation[10];
        scanf("%s", operation);
        
        if (strcmp(operation, "insert") == 0) {
            int value;
            scanf("%d", &value);
            insert(&heap, value);
        } else if (strcmp(operation, "extract") == 0) {
            int result;
            if (extractMin(&heap, &result)) {
                printf("%d\n", result);
            } else {
                printf("Heap is empty\n");
            }
        } else if (strcmp(operation, "peek") == 0) {
            int result;
            if (peek(&heap, &result)) {
                printf("%d\n", result);
            } else {
                printf("Heap is empty\n");
            }
        } else if (strcmp(operation, "empty") == 0) {
            printf("%s\n", isEmpty(&heap) ? "true" : "false");
        }
    }
    
    return 0;
}
